### PR TITLE
[gemma3] fix bidirectional attention mask

### LINKED ISSUE #38053: Attention mask for multi-image input in gemma3
### System Info

As per the attention mask example in the Gemma3 blog (https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/gemma3/attention-ascii.png), it looks like there is non-causal attention within the image and causal attention across images (i.e., an image does not attend to a future image). However, when running gemma3 generate using transformers (v4.51.3), looks like there is non-causal attention across images. 

```
import torch
from transformers import AutoProcessor, Gemma3ForConditionalGeneration
import os
import pickle
import torch._dynamo
torch._dynamo.config.suppress_errors = True

ckpt = "google/gemma-3-4b-it"
model = Gemma3ForConditionalGeneration.from_pretrained(
    ckpt, device_map="auto", torch_dtype=torch.float32,
)
processor = AutoProcessor.from_pretrained(ckpt)

messages = [
    {
        "role": "user",
        "content": [
            {"type": "text", "text": "First image: "},
            {"type": "image", "path": "img1.jpg"},
            {"type": "text", "text": "Second image:"},
            {"type": "image", "path": "img2.jpg"},
            {"type": "text", "text": "Describe all images in single sentence."}
        ]
    }
]
inputs = processor.apply_chat_template(
    messages, add_generation_prompt=True, tokenize=True,
    return_dict=True, return_tensors="pt"
).to(model.device)

generation = model.generate(**inputs, max_new_tokens=1, return_dict_in_generate=True,output_attentions=True , do_sample=False)
```

Image

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Code attached

### Expected behavior

Should the attention across image be non-causal or causal?

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/38053. Images cannot attend to future images but in current implementation we unmask attention within all images

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
[run-slow] gemma3

**Comment by zucchini-nlp** [comment]
run-slow: gemma3

**Comment by zucchini-nlp** [comment]
image


What I got with the current state, and a few more tweaks for gemma3 locally (for the weird image token 🙃 )

File: src/transformers/models/gemma3/modeling_gemma3.py [hunk 1]
```diff
@@ -1063,9 +1063,23 @@ def _update_causal_mask(
...
+            padded = nn.functional.pad(is_image, (1, 0), value=0)
+            new_image_start = is_image & ~padded[:, :-1]
+            image_group_id = torch.cumsum(new_image_start.int(), dim=1) - 1
+            image_group_positions = torch.where(is_image, image_group_id, torch.full_like(token_type_ids, -1))
+            same_image_mask = (image_group_positions.unsqueeze(1) == image_group_positions.unsqueeze(2)) & (
+                image_group_positions.unsqueeze(1) != -1
+            )
+            same_image_mask = same_image_mask.unsqueeze(1)
+
+            image_mask = token_type_mask & same_image_mask
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/gemma3/modeling_gemma3.py:1079]: oke, verified that it is correct now, will refactor with bucketize next

**Review by ArthurZucker — APPROVED** [review]
Thanks it looks better!
